test/pytest/test_bridge.py

```diff
@@ -655,6 +677,117 @@ async def test_fsreplace1_error(transport: MockTransport, tmp_path: Path) -> Non
+        with unittest.mock.patch('grp.getgrnam', side_effect=KeyError()):
+            await transport.check_open('fsreplace1', path=str(tmp_path / 'test'),
+                                       attrs={'user': 'bazinga', 'group': 'test'},
+                                       problem='not-found',
+                                       reply_keys={
+                                           'message': 'gid not found for test'
+            })
+
+
+ATTRS_TEST_DATA = [
+    (GrpPwMock(1111, 1110), {'user': 1111, 'group': 1110}),
+    (GrpPwMock(1111, 1110), {'user': 'monkey', 'group': 'group'}),
+]
+
+
+@pytest.mark.parametrize(('user_group_mock_arg', 'attrs'), ATTRS_TEST_DATA)
+@pytest.mark.asyncio
+async def test_fsreplace1_attrs(transport: MockTransport, fchown_mock: unittest.mock.MagicMock,
+                                tmp_path: Path, user_group_mock, attrs: Dict[str, Union[int, str]]) -> None:
+    async def existing_file_with_attrs(filename: str, tag) -> None:
+        test_file = tmp_path / filename
+        ch = await transport.check_open('fsreplace1', path=str(test_file), attrs=attrs, tag=tag)
+        transport.send_data(ch, b'content')
+        transport.send_done(ch)
+        await transport.assert_msg('', command='done', channel=ch)
+        await transport.check_close(ch)
+        assert test_file.read_bytes() == b'content'
+
+    async def create_file_with_attrs(filename: str) -> None:
+        test_file = tmp_path / filename
+        ch = await transport.check_open('fsreplace1', path=str(test_file), attrs=attrs)
+        transport.send_data(ch, b'content')
+        transport.send_done(ch)
+        await transport.assert_msg('', command='done', channel=ch)
+        await transport.check_close(ch)
+        assert test_file.read_bytes() == b'content'
+
+    await create_file_with_attrs('test')
+    fchown_mock.assert_called()
+    _, call_uid, call_gid = fchown_mock.call_args[0]
+    assert call_uid == user_group_mock.pw_uid
+    if user_group_mock.gr_gid is None:
+        assert call_gid == user_group_mock.pw_uid
+    else:
+        assert call_gid == user_group_mock.gr_gid
+
+    # Existing file
+    existing_file = tmp_path / 'existing'
+    existing_file.write_text('new')
+
+    # Obtain tag
+    ch = await transport.check_open('fsread1', path=str(existing_file))
+    transport.send_done(ch)
+    await transport.assert_data(ch, b'new')
+    await transport.assert_msg('', command='done', channel=ch)
+    transport.send_close(ch)
+    close_msg = await transport.next_msg('')
+    assert close_msg['command'] == 'close'
+    tag = close_msg['tag']
+
+    await existing_file_with_attrs('existing', tag)
+    fchown_mock.assert_called()
+    _, call_uid, call_gid = fchown_mock.call_args[0]
+    assert call_uid == user_group_mock.pw_uid
+    if user_group_mock.gr_gid is None:
+        assert call_gid == user_group_mock.pw_uid
+    else:
+        assert call_gid == user_group_mock.gr_gid
+
```

Comment:
It would be nice to see tests for all of the combinations:
 - replace existing file without specifying tag or attrs
 - replace existing file with tag but no attrs
 - replace existing file with no attrs, but tag
 - replace existing file with attrs and tag (this is the case you test here, if I read it correctly)